src/amplitude_experiment/deployment/deployment_runner.py

```diff
@@ -0,0 +1,113 @@
+import logging
+from typing import Optional, Set
+import threading
+
+from ..exception import CohortNotModifiedException
+from ..local.config import LocalEvaluationConfig
+from ..cohort.cohort_loader import CohortLoader
+from ..cohort.cohort_storage import CohortStorage
+from ..flag.flag_config_api import FlagConfigApi
+from ..flag.flag_config_storage import FlagConfigStorage
+from ..local.poller import Poller
+from ..util.flag_config import get_all_cohort_ids_from_flag
+
+
+class DeploymentRunner:
+    def __init__(
+            self,
+            config: LocalEvaluationConfig,
+            flag_config_api: FlagConfigApi,
+            flag_config_storage: FlagConfigStorage,
+            cohort_storage: CohortStorage,
+            cohort_loader: Optional[CohortLoader] = None,
+    ):
+        self.config = config
+        self.flag_config_api = flag_config_api
+        self.flag_config_storage = flag_config_storage
+        self.cohort_storage = cohort_storage
+        self.cohort_loader = cohort_loader
+        self.lock = threading.Lock()
+        self.poller = Poller(self.config.flag_config_polling_interval_millis / 1000, self.__periodic_refresh)
+        self.logger = logging.getLogger("Amplitude")
+        self.logger.addHandler(logging.StreamHandler())
+        if self.config.debug:
+            self.logger.setLevel(logging.DEBUG)
+
+    def start(self):
+        with self.lock:
+            self.refresh()
+            self.poller.start()
+
+    def stop(self):
+        self.poller.stop()
+
+    def __periodic_refresh(self):
+        try:
+            self.refresh()
+        except Exception as e:
+            self.logger.error(f"Refresh flag and cohort configs failed: {e}")
+
+    def refresh(self):
+        self.logger.debug("Refreshing flag configs.")
+        try:
+            flag_configs = self.flag_config_api.get_flag_configs()
+        except Exception as e:
+            self.logger.error(f'Failed to fetch flag configs: {e}')
+            raise Exception
+
+        flag_keys = {flag['key'] for flag in flag_configs}
+        self.flag_config_storage.remove_if(lambda f: f['key'] not in flag_keys)
+
+        for flag_config in flag_configs:
+            cohort_ids = get_all_cohort_ids_from_flag(flag_config)
+            if not self.cohort_loader or not cohort_ids:
+                self.logger.debug(f"Putting non-cohort flag {flag_config['key']}")
+                self.flag_config_storage.put_flag_config(flag_config)
+                continue
+
+            # Keep track of old flag and cohort for each flag
+            old_flag_config = self.flag_config_storage.get_flag_config(flag_config['key'])
+
+            try:
+                self._load_cohorts(flag_config, cohort_ids)
```

Comment:
Assuming two flags are using the same cohort, if first cohort download success, the second flag would be using the new cohort while the second flag itself being out of date.